eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/UtxowSpec.hs

```diff
@@ -49,3 +61,69 @@ spec = describe "UTXOW" $ do
             mkBasicTx txBody
               & (witsTxL %~ (bootAddrTxWitsL .~ [aliceBadWitness]))
       submitFailingTx txBad [injectFailure $ InvalidWitnessesUTXOW [aliceVKey]]
+
+  it "MissingVKeyWitnessesUTXOW" $ do
+    aliceKh <- freshKeyHash
+    txIn <- sendCoinTo (Addr Testnet (KeyHashObj aliceKh) StakeRefNull) (Coin 99)
+    withPostFixup (\tx -> pure (tx & bodyTxL . inputsTxBodyL <>~ [txIn])) $
+      submitFailingTxWithSubset
+        (mkBasicTx mkBasicTxBody)
+        [ injectFailure $
+            MissingVKeyWitnessesUTXOW [coerceKeyRole aliceKh]
+        ]
+
+  it "MissingScriptWitnessesUTXOW" $ do
+    requiredKh <- freshKeyHash
+    let scriptHash = hashScript @era $ fromNativeScript $ RequireSignature requiredKh
+    txIn <- produceScript scriptHash
+    withPostFixup (\tx -> pure (tx & bodyTxL . inputsTxBodyL <>~ [txIn])) $
+      submitFailingTxWithSubset
+        (mkBasicTx mkBasicTxBody)
+        [ injectFailure $
+            MissingScriptWitnessesUTXOW [scriptHash]
+        ]
+
+  it "MissingTxBodyMetadataHash" $ do
+    auxData <- arbitrary @(TxAuxData era)
+    let auxDataHash = hashTxAuxData auxData
+    let tx = mkBasicTx mkBasicTxBody & auxDataTxL .~ SJust auxData
+    submitFailingTx
+      tx
+      [ injectFailure $
+          MissingTxBodyMetadataHash auxDataHash
+      ]
+
+  it "MissingTxMetadata" $ do
+    auxData <- arbitrary @(TxAuxData era)
+    let auxDataHash = hashTxAuxData auxData
+    let tx = mkBasicTx mkBasicTxBody & bodyTxL . auxDataHashTxBodyL .~ SJust auxDataHash
+    submitFailingTx
+      tx
+      [ injectFailure $
+          MissingTxMetadata auxDataHash
+      ]
+
+  it "ConflictingMetadataHash" $ do
+    auxData <- arbitrary @(TxAuxData era)
+    let auxDataHash = hashTxAuxData auxData
+    wrongAuxDataHash <- arbitrary @(AuxiliaryDataHash (EraCrypto era))
+    let tx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . auxDataHashTxBodyL .~ SJust wrongAuxDataHash
+            & auxDataTxL .~ SJust auxData
+    submitFailingTx
+      tx
+      [ injectFailure $
+          ConflictingMetadataHash wrongAuxDataHash auxDataHash
+      ]
+
+  it "ExtraneousScriptWitnessesUTXOW" $ do
+    requiredKh <- freshKeyHash
+    let script = fromNativeScript $ RequireSignature requiredKh
+    let scriptHash = hashScript @era script
+    let tx = mkBasicTx mkBasicTxBody & witsTxL . scriptTxWitsL .~ [(scriptHash, script)]
+    submitFailingTxWithSubset
+      tx
+      [ injectFailure $
+          ExtraneousScriptWitnessesUTXOW [scriptHash]
+      ]
```

Comment:
This is a more accurate way of testing predicate failures that also nicely depict changes between eras:
```suggestion
        -- We dropped validating scripts when they are not needed starting with Babbage
        expectedFailures
          | eraProtVerLow @era >= natVersion @6 =
              [ injectFailure $ ExtraneousScriptWitnessesUTXOW [scriptHash]
              ]
          | otherwise =
              [ injectFailure $ ScriptWitnessNotValidatingUTXOW [scriptHash]
              , injectFailure $ ExtraneousScriptWitnessesUTXOW [scriptHash]
              ]
    submitFailingTx tx expectedFailures
```